rosbag2_transport/test/rosbag2_transport/test_topic_filter.cpp

```diff
@@ -58,6 +61,7 @@ TEST(TestTopicFilter, filter_hidden_topics) {
   }
   {
     rosbag2_transport::RecordOptions record_options;
+    record_options.all_topics = true;
```

Comment:
@MichaelOrlov 

It is related to the logic in TopicFilter::take_topic().  
```c++
// Original code
struct RecordOptions
{
public:
  bool all = false;
  bool is_discovery_disabled = false;
  std::vector<std::string> topics;
  std::string rmw_serialization_format;
  std::chrono::milliseconds topic_polling_interval{100};
  std::string regex = "";
  std::string exclude = "";
  std::string node_prefix = "";
  std::string compression_mode = "";
  std::string compression_format = "";
  uint64_t compression_queue_size = 1;
  uint64_t compression_threads = 0;
  std::unordered_map<std::string, rclcpp::QoS> topic_qos_profile_overrides{};
  bool include_hidden_topics = false;
  bool include_unpublished_topics = false;
  bool ignore_leaf_topics = false;
  bool start_paused = false;
  bool use_sim_time = false;
};
```
For original code, if `all==false`, `topics is empty` and `regex is empty`, TopicFilter::take_topic() can return true. It is because parameter check in record.py can avoid this case.  Original test code took advantage of this 'vulnerability'. But current logic in TopicFilter::take_topic()  cannot allow this way. So I have to add code `all_topics = true`.

Reply:
> On one hand, using the generated bag files in tests is a working approach, however on the other hand it will be difficult to support such test infrastructure in the future.
Sometimes we make changes in the bag format of metadata and have to regenerate all bags written for tests.
It would be better to write bags on the fly from each test to the temporary folder.
Could you please change tests this way?
You can find an example of how to write test bags on the fly in the other tests in the rosbag2_cpp package. For instance this one.

Yes. I didn't consider this case. I'll update the test code to remove the pre-prepared bag files. 
Thanks for your advice.

Reply:
@MichaelOrlov 

> On one hand, using the generated bag files in tests is a working approach, however on the other hand it will be difficult to support such test infrastructure in the future.
Sometimes we make changes in the bag format of metadata and have to regenerate all bags written for tests.
It would be better to write bags on the fly from each test to the temporary folder.
Could you please change tests this way?
You can find an example of how to write test bags on the fly in the other tests in the rosbag2_cpp package. For instance this one.  

For service event message, recording action cannot be made like topic message
```
    test_msgs::msg::BasicTypes msg;
    writer.write(msg, "testtopic", rclcpp::Time{});
```
So test code have to actually record the service event messages. These tests cannot be located in rosbag2_cpp. So I move them to rosbag2_tests.  
Commit https://github.com/ros2/rosbag2/pull/1480/commits/872e72b358067079828cb38069f169dda03ddaa7 implemented it.

Reply:
> For original code, if `all==false`, `topics is empty` and `regex is empty`, TopicFilter::take_topic() can return true. It is because parameter check in record.py can avoid this case. Original test code took advantage of this 'vulnerability'. But current logic in TopicFilter::take_topic() cannot allow this way. So I have to add code `all_topics = true`.

BTW according to the latest logic in the `take_topic()` we allow to take topic hidden topics  if `all==false`, `topics is empty` and `regex is empty` and record_options_.include_hidden_topics==true;
```
 if (!is_service_event_topic) {
    if (!record_options_.all_topics &&
      record_options_.topics.empty() &&
      record_options_.regex.empty() &&
      !record_options_.include_hidden_topics)
    {
      return false;
    }
```
It looks like we either should remove the `record_options.all_topics = true;` from the  `TestTopicFilter::filter_hidden_topics` or change logic in the `take_topic` method.
Also make sense to remove `record_options.all_topics = true;` from the  `TestTopicFilter::filter_topics_with_more_than_one_type` this way we will verify that condition for filtering more than one type are at the begging of the checks.

Reply:
Yes. `all_topics = true` should be removed. Otherwise, related codes aren't tested.  
I will update codes.

Reply:
@MichaelOrlov 

The following check didn't allow that all_topics==false, topics is empty and regex is empty at the same.

https://github.com/Barry-Xu-2018/rosbag2/blob/872e72b358067079828cb38069f169dda03ddaa7/rosbag2_transport/src/rosbag2_transport/topic_filter.cpp#L161-L174
```c++
    if (!record_options_.all_topics) {
      // Not in include topic list
      if (record_options_.topics.empty() || !topic_in_list(topic_name, record_options_.topics)) {
        // Not match include regex
        if (!record_options_.regex.empty()) {
          std::regex include_regex(record_options_.regex);
          if (!std::regex_search(topic_name, include_regex)) {
            return false;
          }
        } else {
          return false;
        }
      }
    }
```
For topic, the following one of condition must be matched while record_options_.include_hidden_topics==true. Other no topic can pass.  
- record_options_.all_topics is true
- record_options_.topics isn't empty
- record_options_.regex isn't empty

So `record_options.all_topics = true` is needed in tests.